### PR TITLE
FAD-6066 Setup Sending Domain for sending section

### PR DESCRIPTION
**Pull request by gwalls**

### CODE REVIEW

File: src/actions/sendingDomains.js [hunk 1]
```diff
@@ -19,3 +19,26 @@ export function get(id) {
     }
   });
 }
+
+export function verifyDNS(id, type) {
```
Comment (rajumsys) [inline src/actions/sendingDomains.js]: fyi, we both have this implementation with different names

File: src/config/default.js [hunk 2]
```diff
@@ -30,6 +30,9 @@ const config = {
     supportEmail: 'support@sparkpost.com',
     billingEmail: 'billing@sparkpost.com'
   },
+  featureFlags: {
```
Comment (rajumsys) [inline src/config/default.js]: i guess we can simply create one AE/AD PR for all this changes (vs one for each section). what do you think? my ticket FAD-6064 needs to add bounce domain info.

File: src/pages/sendingDomains/components/SetupSending.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      <Icon name="Error"/>
+    );
+  }
+
+  renderTxtRecordPanel() {
+    const { domain, id } = this.props;
+
+    return (
+      <Panel sectioned>
+        <Table>
```
Comment (rajumsys) [inline src/pages/sendingDomains/components/SetupSending.js]: I've a `SimpleTable` component in my PR. You can re-use it if you want. Either add those styles there or need to configure them to accept those class names.

File: src/reducers/sendingDomains.js [hunk 4]
```diff
@@ -21,6 +21,15 @@ export default (state = initialState, { type, payload }) => {
     case 'GET_SENDING_DOMAIN_FAIL':
       return { ...state, getError: payload, getLoading: false };
 
+    case 'VERIFY_SENDING_DOMAIN_PENDING':
+      return { ...state, verifyLoading: true, verifyError: null };
+
+    case 'VERIFY_SENDING_DOMAIN_SUCCESS':
+      return { ...state, domain: { ...state.domain, status: payload }, verifyLoading: false };
```
Comment (rajumsys) [inline src/reducers/sendingDomains.js]: i've corrected this in my PR
```      return { ...state, verifyLoading: false, domain: { ...state.domain, status: payload }};
```

File: src/config/default.js (hunk 2 above)
Comment (gwalls) [reply src/config/default.js]: Sure that makes sense to me. I can go ahead and do that.

File: src/pages/sendingDomains/components/SetupSending.js (hunk 3 above)
Comment (gwalls) [reply src/pages/sendingDomains/components/SetupSending.js]: Cool, that'll look much cleaner if I can get it to work.

File: src/reducers/sendingDomains.js (hunk 4 above)
Comment (gwalls) [reply src/reducers/sendingDomains.js]: Just change the order? Will do

File: src/actions/sendingDomains.js (hunk 1 above)
Comment (gwalls) [reply src/actions/sendingDomains.js]: Okay, I can change the name to match yours.

File: src/reducers/sendingDomains.js (hunk 4 above)
Comment (rajumsys) [reply src/reducers/sendingDomains.js]: sorry my bad, i misread. order doesn't matter in this case! your code is fine.

File: src/pages/sendingDomains/EditPage.js [hunk 5]
```diff
@@ -22,6 +26,18 @@ export class EditPage extends Component {
...
 
+  renderSections = () => {
+    const { domain, match: { params: { id }}} = this.props;
+
+    if (!domain) {
+      return null;
+    }
+
+    return (
+      <SetupSending form={FORM_NAME} id={id} domain={domain}/>
```
Comment (bizob2828) [inline src/pages/sendingDomains/EditPage.js]: is there a reason why this form is getting passed in and not just set on the component? reason i ask is i've been looking at the other PRs for this page to get a sense of how aligned I am and this caught me off guard

File: src/actions/sendingDomains.js [hunk 6]
```diff
@@ -19,3 +19,27 @@ export function get(id) {
     }
   });
 }
+
+export function verify(id, type) {
```
Comment (bizob2828) [inline src/actions/sendingDomains.js]: sending domains are sending domain aware, i think we need to include those headers

Comment (jasonrhodes) [reply src/actions/sendingDomains.js]: You mean "subaccount-aware"? They don't seem to be based on what we've seen in the existing UI where we don't include the subaccount id in a query param at all for the edit page.

Comment (rajumsys) [reply src/actions/sendingDomains.js]: ha ha, i was scratching my head what that meant!

Comment (bizob2828) [reply src/actions/sendingDomains.js]: @jasonrhodes interesting, ok not sure how it works then

File: src/actions/sendingDomains.js [hunk 7]
```diff
@@ -19,3 +19,27 @@ export function get(id) {
...
+    type: 'VERIFY_SENDING_DOMAIN',
+    meta: {
+      method: 'POST',
+      url: `/sending-domains/${id}/verify`,
+      data
+    }
+  });
+}
+
+export function update(id, data) {
```
Comment (bizob2828) [inline src/actions/sendingDomains.js]: need set subaccount header here too.

File: src/actions/sendingDomains.js (hunk 6 above)
Comment (bizob2828) [reply src/actions/sendingDomains.js]: here's what i did in update ```
--- a/src/actions/sendingDomains.js
+++ b/src/actions/sendingDomains.js
@@ -1,4 +1,5 @@
 import sparkpostApiRequest from 'src/actions/helpers/sparkpostApiRequest';
+import setSubaccountHeader from './helpers/setSubaccountHeader';

 export function list() {
   return sparkpostApiRequest({
@@ -20,14 +21,15 @@ export function get(id) {
   });
 }

-// TODO: wire up for subaccount support
-export function update(id, data) {
+export function update({ sendingDomain, subaccount, ...data }) {
+  const headers = setSubaccountHeader(subaccount);
   return sparkpostApiRequest({
     type: 'UPDATE_SENDING_DOMAIN',
     meta: {
       method: 'PUT',
-      url: `/sending-domains/${id}`,
-      data
+      url: `/sending-domains/${sendingDomain}`,
+      data,
+      headers
     }
   });
 }
```

File: src/actions/sendingDomains.js (hunk 7 above)
Comment (gwalls) [reply src/actions/sendingDomains.js]: I'm just going to remove this as I didn't have to use it

File: src/pages/sendingDomains/EditPage.js (hunk 5 above)
Comment (gwalls) [reply src/pages/sendingDomains/EditPage.js]: nixing it

File: src/actions/sendingDomains.js (hunk 6 above)
Comment (gwalls) [reply src/actions/sendingDomains.js]: fwiw, both the update and verify include subaccount in the Angular UI `src/common/resources/sending-domains.js`
```  _postToVerify(domain, data, subaccount) {
    let url = `${this.apiBase}/${domain}/verify`
      , opts = this.SubaccountHeader.set(subaccount);

    return this.Http.promisify(this.$http.post(url, data, opts));
  }```

File: src/pages/sendingDomains/components/SetupSending.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    } else if (!readyFor.sending && !readyFor.dkim && config.featureFlags.allow_mailbox_verification) {
+      return (<p>We recommend DNS verification, but if you don't have DNS access, you can <a>set this domain up for sending via email</a>.</p>);
+    } else if (readyFor.sending && !readyFor.dkim) {
+      return (
+        <Banner status="warning">
+          This domain is "ready for sending" but if you plan to use it for sending, we still recommend that you add this TXT record to make it "ready for DKIM" as well.
+        </Banner>
+      );
+    }
+    // TODO: what about (!readyFor.sending && readyFor.dkim)?
```
Comment (gwalls) [inline src/pages/sendingDomains/components/SetupSending.js]: should probably just 'else return null', if nothing else

**Comment by gwalls** [comment]
changes for config: https://fisheye.int.messagesystems.com/cru/APP-3120

File: src/pages/sendingDomains/components/SetupSending.js (hunk 8 above)
Comment (jasonrhodes) [reply src/pages/sendingDomains/components/SetupSending.js]: That specific situation in that TODO can't happen—being ready for DKIM means you are necessarily ready for sending, at least as it is today.

(But agreed with Greg that return null at the bottom (else or no else) is good.)

File: src/pages/sendingDomains/components/SetupSending.js (hunk 3 above)
Comment (bizob2828) [reply src/pages/sendingDomains/components/SetupSending.js]: either way i'd move this into a diff component, whether that's SimpleTable or something specific for this

File: src/pages/sendingDomains/components/SendingDomainSection.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
```
Comment (bizob2828) [inline src/pages/sendingDomains/components/SendingDomainSection.js]: missing tests for this but i wrote them in my PR so don't worry

File: src/pages/sendingDomains/components/tests/SetupSending.test.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        status
+      },
+      verify: jest.fn(() => Promise.resolve()),
+      showAlert: jest.fn()
+    };
+
+    config.featureFlags.allow_mailbox_verification = true;
+    wrapper = shallow(<SetupSending {...props}/>);
+  });
+
```
Comment (bizob2828) [inline src/pages/sendingDomains/components/tests/SetupSending.test.js]: missing a test to set verifyLoading props and snapshot the button differences

File: src/config/default.js (hunk 2 above)
Comment (gwalls) [reply src/config/default.js]: I could change this to camelCase, but I'd no longer be able to pipe `featureFlags` to `to_nice_json` in ansible.

File: src/pages/sendingDomains/components/SendingDomainSection.js (hunk 9 above)
Comment (gwalls) [reply src/pages/sendingDomains/components/SendingDomainSection.js]: I was counting on that 😄

File: src/pages/sendingDomains/components/SetupSending.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        : <Button plain onClick={this.verifyDomain}>Verify TXT Record</Button>
+    );
+  }
+
+  renderIcon() {
+    const { domain } = this.props;
+    const readyFor = resolveReadyFor(domain.status);
+
+    return (
+      readyFor.dkim
```
[the remark this answers is not in the document]
Comment (gwalls) [reply src/pages/sendingDomains/components/SetupSending.js]: Per his recommendation, I'm going to move the the colors into a css class

File: src/pages/sendingDomains/components/tests/SetupSending.test.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    beforeEach(() => {
+      instance = wrapper.instance();
+    });
+
+    it('verifies domain and alerts when verification successful', async() => {
+      props.verify.mockReturnValue(Promise.resolve({ dkim_status: 'valid' }));
+      await instance.verifyDomain();
+      expect(props.verify).toHaveBeenCalledTimes(1);
+      expect(props.verify).toHaveBeenCalledWith('xyz.com', 'dkim', 999);
+      const arg = props.showAlert.mock.calls[0][0];
```
[the remark this answers is not in the document]
Comment (gwalls) [reply src/pages/sendingDomains/components/tests/SetupSending.test.js]: Hmmm the snapshot wasn't picking up the Alert component

**Comment by bizob2828** [comment]
merged into FAD-6065

**Pull request closed without merging**